resources/lemmas.md

```diff
@@ -184,12 +183,19 @@ The rules are applied only when the side-conditions are met.
 These rules are specific to reasoning about EVM programs.
 
 ```k
-    rule (I1 +Int I2) +Int I3 => I1 +Int (I2 +Int I3) when #isConcrete(I2) andBool #isConcrete(I3)
-    rule (I1 +Int I2) -Int I3 => I1 +Int (I2 -Int I3) when #isConcrete(I2) andBool #isConcrete(I3)
-    rule (I1 -Int I2) +Int I3 => I1 -Int (I2 -Int I3) when #isConcrete(I2) andBool #isConcrete(I3)
-    rule (I1 -Int I2) -Int I3 => I1 -Int (I2 +Int I3) when #isConcrete(I2) andBool #isConcrete(I3)
+    //orienting symbolic term to be first, converting -Int to +Int
+    rule I +Int B => B          +Int I when #isConcrete(I) andBool notBool #isConcrete(B)
+    rule I -Int B => (0 -Int B) +Int I when #isConcrete(I) andBool notBool #isConcrete(B) andBool I =/=Int 0
+    rule A -Int I => A +Int (0 -Int I) when notBool #isConcrete(A) andBool #isConcrete(I)
```

Comment:
it turned out that the above two rules are two strong (applicable in too many contexts).  every time we have some special form of symbolic terms, say `foo(X) - Y`, that we want to have its own simplification rule, we need to have either add more side-condition to the above rules to prevent them from being applied for the special terms, or we need to write a simplification rule for the new term that results from applying the above rule, i.e., `foo(X) + (-Y)`.

i'm wondering if the above two rules are really needed to fix the ambiguity, or you generalized all the way up, potentially being more than what is necessary, as you went through.

Reply:
I strongly propose not to have the normalization from -Int to +Int, unless it is necessary.

Or, if it is really necessary, I think we may be able to drop many of these arithmetic simplification rules, since they came from simplifying the gas terms, but now we are using a new gas abstraction which we can control more.

What do you think guys? @denis-bogdanas @yzhang90

Reply:
@daejunpark Ok, I removed -Int normalization for terms like `I -Int A`, but I kept it for `A -Int I`. Hope Jenkins is fine with that.

If I remove -Int normalziation entirely, but keep +Int reordering rule, then I get this mess:
```
[Error] Critical: [non-deterministic function definition]: more than one rule
can apply to the term:
_+Int_(
  Int(#"0"),,
  _-Int__INT(
    _*Int__INT(Int(#"65"),, NR_SIG_1903:Int),,
    Int(#"32")
  )
)

Candidate rules:
    rule I +Int B => B          +Int I when #isConcrete(I) andBool notBool
#isConcrete(B)
	Source: /mnt/d/verified-smart-contracts-fork/specs/lemmas.k
Location(187,10,187,90)

    rule I1 +Int (B -Int I3) => B +Int (I1 -Int I3) when #isConcrete(I1)
andBool notBool #isConcrete(B) andBool #isConcrete(I3)
	Source: /mnt/d/verified-smart-contracts-fork/specs/lemmas.k
Location(198,10,198,128)

Candidate results:
_-Int__INT(
  _*Int__INT(Int(#"65"),, NR_SIG_1903:Int),,
  Int(#"32")
)

_+Int_(
  _*Int__INT(Int(#"65"),, NR_SIG_1903:Int),,0
  Int(#"-32")
)
```

This will disappear if I also remove +Int reordering, but then we'll get a ton of issues elsewhere when 2 rules eventually rewrite something to a sum of terms but in a different order. So the current solution should be a good compromise.

If you still disagree please msg me on slack so that we can reach a consensus quickly.

Reply:
Jenkins results: now proxied-token fails. All gnosis projects pass. With previous version of lemmas all projects were passing. I disabled proxied-token to avoid spending time on this.

Reply:
@daejunpark @yzhang90 Actually a clean normalization solution would be to move symbolic terms of a sum to the RIGHT, not to the left:
```
rule A +Int I => I          +Int A when #isConcrete(I) andBool notBool #isConcrete(B)
```
That's because  both +B and -B can stay at the right-end of a sum, but only +B can stay at the left end. Damn it. It would have all advantages of my initial solution and no disadvantages. Should i do this change?